File: appium_tests/pages/base_page.py
"""
base_page.py — Base Page Object with common element interactions
"""
import time
from appium.webdriver.common.appiumby import AppiumBy
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import (
    TimeoutException, NoSuchElementException, WebDriverException
)
from config import EXPLICIT_WAIT, ELEMENT_WAIT


class BasePage:
    """All Page Objects inherit from this class."""

    def __init__(self, driver):
        self.driver = driver
        self.wait = WebDriverWait(driver, EXPLICIT_WAIT)
        self.short_wait = WebDriverWait(driver, ELEMENT_WAIT)

    # ── Wait helpers ──────────────────────────────────────────────

    def wait_for_element(self, by, value, timeout=None):
        t = timeout or ELEMENT_WAIT
        try:
            return WebDriverWait(self.driver, t).until(
                EC.presence_of_element_located((by, value))
            )
        except TimeoutException:
            return None

    def wait_for_clickable(self, by, value, timeout=None):
        t = timeout or ELEMENT_WAIT
        try:
            return WebDriverWait(self.driver, t).until(
                EC.element_to_be_clickable((by, value))
            )
        except TimeoutException:
            return None

    # ── Find helpers ──────────────────────────────────────────────

    def find(self, by, value):
        try:
            return self.driver.find_element(by, value)
        except (NoSuchElementException, WebDriverException):
            return None

    def finds(self, by, value):
        try:
            return self.driver.find_elements(by, value)
        except (NoSuchElementException, WebDriverException):
            return []

    def find_by_text(self, text):
        return self.find(
            AppiumBy.ANDROID_UIAUTOMATOR,
            f'new UiSelector().text("{text}")'
        )

    def find_by_partial_text(self, text):
        return self.find(
            AppiumBy.ANDROID_UIAUTOMATOR,
            f'new UiSelector().textContains("{text}")'
        )

    def find_field_by_hint(self, hint_text):
        return self.find(
            AppiumBy.ANDROID_UIAUTOMATOR,
            f'new UiSelector().text("{hint_text}")'
        )

    # ── Interaction helpers ───────────────────────────────────────

    def tap(self, by, value, timeout=None):
        el = self.wait_for_clickable(by, value, timeout)
        if el:
            el.click()
            return True
        return False

    def type_text(self, by, value, text, clear=True):
        el = self.wait_for_element(by, value)
        if el:
            if clear:
                el.clear()
            el.send_keys(text)
            return True
        return False

    def type_into_hint(self, hint_text, text):
        """Type into a Flutter TextField identified by its hint."""
        el = self.find_field_by_hint(hint_text)
        if el:
            el.clear()
            el.send_keys(text)
            return True
        return False

    def is_displayed(self, by, value, timeout=5):
        el = self.wait_for_element(by, value, timeout=timeout)
        try:
            return el is not None and el.is_displayed()
        except Exception:
            return False

    def is_text_present(self, text, timeout=5):
        el = self.wait_for_element(
            AppiumBy.ANDROID_UIAUTOMATOR,
            f'new UiSelector().textContains("{text}")',
            timeout=timeout
        )
        return el is not None

    def get_text(self, by, value):
        el = self.find(by, value)
        return el.text if el else ""

    # ── Navigation ────────────────────────────────────────────────

    def back(self):
        try:
            self.driver.back()
        except Exception:
            pass

    def hide_keyboard(self):
        try:
            self.driver.hide_keyboard()
        except Exception:
            pass

    # ── Swipe ─────────────────────────────────────────────────────

    def swipe_up(self):
        size = self.driver.get_window_size()
        w, h = size["width"], size["height"]
        self.driver.swipe(w // 2, int(h * 0.75), w // 2, int(h * 0.25), 600)

    def swipe_down(self):
        size = self.driver.get_window_size()
        w, h = size["width"], size["height"]
        self.driver.swipe(w // 2, int(h * 0.25), w // 2, int(h * 0.75), 600)

    # ── Utility ───────────────────────────────────────────────────

    def sleep(self, seconds=1):
        time.sleep(seconds)

    def get_page_source(self):
        try:
            return self.driver.page_source
        except Exception:
            return ""

    def is_keyboard_shown(self):
        try:
            return self.driver.is_keyboard_shown()
        except Exception:
            return False

File: automation/src/main/java/com/smartstudent/pages/BasePage.java
package com.smartstudent.pages;

import com.smartstudent.drivers.DriverManager;
import com.smartstudent.utils.ScreenshotUtils;
import com.smartstudent.utils.WaitUtils;
import io.appium.java_client.android.AndroidDriver;
import io.appium.java_client.pagefactory.AppiumFieldDecorator;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.PageFactory;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;
import java.util.List;

/**
 * Base Page Object providing common actions for all page objects.
 */
public abstract class BasePage {

    protected final Logger logger = LogManager.getLogger(getClass());
    protected AndroidDriver driver;
    protected WaitUtils waitUtils;
    protected ScreenshotUtils screenshotUtils;

    public BasePage() {
        this.driver = DriverManager.getDriver();
        this.waitUtils = new WaitUtils(driver);
        this.screenshotUtils = new ScreenshotUtils(driver);
        PageFactory.initElements(new AppiumFieldDecorator(driver, Duration.ofSeconds(15)), this);
    }

    // ── Click ────────────────────────────────────────────────────

    protected void click(WebElement element) {
        waitUtils.waitForClickable(element);
        element.click();
        logger.debug("Clicked: {}", element);
    }

    protected void click(By locator) {
        WebElement el = waitUtils.waitForVisible(locator);
        el.click();
        logger.debug("Clicked by locator: {}", locator);
    }

    // ── Type ─────────────────────────────────────────────────────

    protected void type(WebElement element, String text) {
        waitUtils.waitForClickable(element);
        element.clear();
        element.sendKeys(text);
        logger.debug("Typed '{}' into element", text);
    }

    protected void type(By locator, String text) {
        WebElement el = waitUtils.waitForVisible(locator);
        el.clear();
        el.sendKeys(text);
    }

    protected void clearAndType(WebElement element, String text) {
        waitUtils.waitForClickable(element);
        element.clear();
        element.sendKeys(text);
    }

    // ── Read ─────────────────────────────────────────────────────

    protected String getText(WebElement element) {
        waitUtils.waitForVisible(element);
        return element.getText().trim();
    }

    protected String getText(By locator) {
        return waitUtils.waitForVisible(locator).getText().trim();
    }

    protected String getAttribute(WebElement element, String attr) {
        waitUtils.waitForVisible(element);
        return element.getAttribute(attr);
    }

    // ── Visibility ───────────────────────────────────────────────

    protected boolean isDisplayed(WebElement element) {
        try {
            return element.isDisplayed();
        } catch (Exception e) {
            return false;
        }
    }

    protected boolean isDisplayed(By locator) {
        try {
            return driver.findElement(locator).isDisplayed();
        } catch (Exception e) {
            return false;
        }
    }

    protected boolean isEnabled(WebElement element) {
        try {
            return element.isEnabled();
        } catch (Exception e) {
            return false;
        }
    }

    // ── Wait Helpers ─────────────────────────────────────────────

    protected WebElement waitForElement(By locator, int timeoutSeconds) {
        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(timeoutSeconds));
        return wait.until(ExpectedConditions.visibilityOfElementLocated(locator));
    }

    protected boolean waitForText(By locator, String text, int timeoutSeconds) {
        try {
            WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(timeoutSeconds));
            return wait.until(ExpectedConditions.textToBePresentInElementLocated(locator, text));
        } catch (Exception e) {
            return false;
        }
    }

    protected boolean waitForElementToDisappear(By locator, int timeoutSeconds) {
        try {
            WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(timeoutSeconds));
            return wait.until(ExpectedConditions.invisibilityOfElementLocated(locator));
        } catch (Exception e) {
            return false;
        }
    }

    // ── List ─────────────────────────────────────────────────────

    protected List<WebElement> getElements(By locator) {
        return driver.findElements(locator);
    }

    protected int getElementCount(By locator) {
        return driver.findElements(locator).size();
    }

    // ── Scroll ───────────────────────────────────────────────────

    protected void scrollToElement(By locator) {
        try {
            driver.findElement(locator).click();
        } catch (Exception e) {
            // Scroll using UiScrollable
            String uiScrollable = "new UiScrollable(new UiSelector().scrollable(true))"
                    + ".scrollIntoView(new UiSelector().resourceId(\"" + locator + "\"))";
            driver.findElement(By.xpath("//*[@scrollable='true']"));
        }
    }

    protected void swipeUp() {
        int height = driver.manage().window().getSize().getHeight();
        int width  = driver.manage().window().getSize().getWidth();
        driver.executeScript("mobile: swipeGesture", new java.util.HashMap<String, Object>() {{
            put("left", width / 4);
            put("top", (int)(height * 0.75));
            put("width", width / 2);
            put("height", (int)(height * 0.5));
            put("direction", "up");
            put("percent", 0.85);
        }});
    }

    protected void swipeDown() {
        int height = driver.manage().window().getSize().getHeight();
        int width  = driver.manage().window().getSize().getWidth();
        driver.executeScript("mobile: swipeGesture", new java.util.HashMap<String, Object>() {{
            put("left", width / 4);
            put("top", (int)(height * 0.25));
            put("width", width / 2);
            put("height", (int)(height * 0.5));
            put("direction", "down");
            put("percent", 0.85);
        }});
    }

    // ── Screenshot ───────────────────────────────────────────────

    protected String captureScreenshot(String name) {
        return screenshotUtils.capture(name);
    }

    // ── Navigate Back ─────────────────────────────────────────────

    protected void navigateBack() {
        driver.navigate().back();
        logger.debug("Navigated back");
    }

    protected void pressHome() {
        driver.executeScript("mobile: pressKey", new java.util.HashMap<String, Object>() {{
            put("keycode", 3);
        }});
    }

    // ── Keyboard ─────────────────────────────────────────────────

    protected void hideKeyboard() {
        try {
            driver.hideKeyboard();
        } catch (Exception e) {
            logger.debug("Keyboard not present or already hidden");
        }
    }

    // ── Abstract ─────────────────────────────────────────────────

    /**
     * Verify the page is loaded/displayed.
     */
    public abstract boolean isPageLoaded();
}
